### PR TITLE
Improve `includes` in directory `core/math/`

### PR DESCRIPTION
**Pull request by Nazarwadim**
In this PR, I changed includes in directory math.

I used forward declaration in some places. 
I also included the files(which have types that the file needs) that are already used in other inclusions. Why: It reduced the connection between files and if we want to rewrite a file to remove an include, we don't have to worry that this include is used by another file.

**Why did I change files not in the math directory?**
This was because the file `math_funcs.h` uses the following inludes.
``` c++
#include "core/math/random_pcg.h"
#include <float.h>
```
They are not used in the `math_funcs.h` file, but are used in several other. I just added the necessary includes to them in order to remove includes from `math_funcs.h`.

**Benchmark**
System info:
Linux Mint 21.2 (Victoria) - X11 - NVIDIA GeForce RTX 3050 Laptop GPU () - 11th Gen Intel(R) Core(TM) i5-11400H @ 2.70GHz (12 Threads)

Compile command:
`scons -j12 optimize=none`

Compilation time in master:
`[Time elapsed: 00:05:30.874]`

Compilation time in this PR:
`[Time elapsed: 00:05:27.627]`

### CODE REVIEW

**Comment by AThousandShips** [comment]
Most of these are redundant, and it doesn't improve compile time by any real amount, at least a few of these should be removed IMO:
* `core/typedefs.h` This is universal and except for some places in core this is guaranteed to be included
* `core/error/error_macros.h` Same for this one
* `core/os/memory.h` Same for this one
* `core/variant/binder_common.h` Internaal to `variant`, shouldn't be included elsewhere
* Any of the `variant`, `string` or `math` types, like `Vector2/3/4(i)`, `Array`, `Dictionary`, `String`, `StringName`, etc. if `variant.h` is included, it implies them very strongly

File: core/math/a_star.h [hunk 1]
```diff
@@ -31,9 +31,20 @@
...
+#include "core/object/object.h"
 #include "core/object/ref_counted.h"
+#include "core/os/memory.h"
+#include "core/string/string_name.h"
+#include "core/templates/hash_set.h"
 #include "core/templates/oa_hash_map.h"
+#include "core/templates/pair.h"
+#include "core/templates/vector.h"
+#include "core/typedefs.h"
+#include "core/variant/variant.h"
```
Comment (AThousandShips) [inline core/math/a_star.h:47]: Here, for example, this doesn't affect anything, as `variant.h` is already included via `object.h`, meaning that it'll just enter the file, hit the include guard, and exit again

**Comment by Nazarwadim** [comment]
I checked it 3 times. Once the time improved by 2 seconds, another time by 7 seconds, and the third time by 1 second. 

I did not delete the cache, just compiled on two different branches. 

This definitely improves the compile time because I removed <float.h> and <random_pcg.h>, which are used in several files, but compiled in almost all files.

Also, re-including a file affects the compilation time by about 0.05% of its total compilation time.  But I think I'm going to take away the includes you mentioned earlier.

**Comment by AThousandShips** [comment]
Did you run the unchanged 3 times as well? But in either case 7 seconds out of over 5 minutes is negligible, it's well within the margins for the operating system handling files

**Comment by Nazarwadim** [comment]
I changed branches every time I tested

**Comment by KoBeWi** [comment]
Changing branches does not mean full recompilation. You could as well change to a differently-named equivalent branch and do a null build 🙃

**Comment by AThousandShips** [comment]
Which is why you should do:
```
scons --clean
scons --cache-disable ...
```
To get a true "from scratch" build

File: core/math/math_funcs.h [hunk 2]
```diff
@@ -33,14 +33,14 @@
 
 #include "core/error/error_macros.h"
 #include "core/math/math_defs.h"
-#include "core/math/random_pcg.h"
 #include "core/typedefs.h"
 
 #include "thirdparty/misc/pcg.h"
 
-#include <float.h>
```
Comment (AThousandShips) [inline core/math/math_funcs.h:41]: So as you can see this change won't work, and is the major change here moving `float.h` out of this file, but that won't compile on windows, which is a good indication of how these kinds of changes have potential fallout and needs to be done carefully

File: core/math/math_funcs.h [hunk 3]
```diff
@@ -33,14 +33,14 @@
 
 #include "core/error/error_macros.h"
 #include "core/math/math_defs.h"
-#include "core/math/random_pcg.h"
```
Comment (AThousandShips) [inline core/math/math_funcs.h:36]: This *also* breaks on windows, which seems to mean that this code needs to be moved to somewhere else:
```cpp
#if defined(__GNUC__)
#define CLZ32(x) __builtin_clz(x)
#elif defined(_MSC_VER)
#include <intrin.h>
static int __bsr_clz32(uint32_t x) {
	unsigned long index;
	_BitScanReverse(&index, x);
	return 31 - index;
}
#define CLZ32(x) __bsr_clz32(x)
#else
#endif
```

**Comment by AThousandShips** [comment]
So after fixing the missing includes on Windows (mind you ran an SCU build so not entirely representative of "fixes all includes"), but from specifically SCU builds this made no difference (granted I didn't remove the `float.h` additions in other files that are redundant if they include `math_funcs.h` as `float.h` is needed in `math_funcs.h`)

Before:
* 00:03:29.191
* 00:03:28.656

After:
* 00:03:27.218
* 00:03:28.320

I'll do a few non-SCU builds too and see but I don't expect any changes there either

File: core/math/math_funcs.h (hunk 3 above)
Comment (Nazarwadim) [reply core/math/math_funcs.h:36]: Well, then, should this solve the errors on Windows?
```c++
#include <float.h>
#ifdef _MSC_VER
#include <intrin.h>
#endif
```

Comment (AThousandShips) [reply core/math/math_funcs.h:36]: It does, but the `float` change removes the main reason here, save for the random part, which isn't very heavy (the third-party code is tiny)

**Comment by AThousandShips** [comment]
Tested a non-scu build and same results, identical time ranges:

Before:
* 00:08:53.617
* 00:08:53.669
* 00:08:58.694

After:
* 00:09:07.868
* 00:08:51.987
* 00:08:54.952

File: core/math/quaternion.cpp [hunk 4]
```diff
@@ -33,6 +33,8 @@
 #include "core/math/basis.h"
 #include "core/string/ustring.h"
 
+#include <math.h>
```
Comment (AThousandShips) [inline core/math/quaternion.cpp:36]: Unnecessary, includes `math_funcs`

File: scene/resources/curve.cpp [hunk 5]
```diff
@@ -32,6 +32,8 @@
 
 #include "core/math/math_funcs.h"
 
+#include <float.h>
```
Comment (AThousandShips) [inline scene/resources/curve.cpp:35]: Unnecessary

File: core/math/vector3.h [hunk 6]
```diff
@@ -32,8 +32,10 @@
 #define VECTOR3_H
 
 #include "core/error/error_macros.h"
+#include "core/math/math_defs.h"
```
Comment (AThousandShips) [inline core/math/vector3.h:35]: All of these `math_defs` includes are unnecessary with `math_funcs`

File: core/math/triangulate.h [hunk 7]
```diff
@@ -31,8 +31,10 @@
 #ifndef TRIANGULATE_H
 #define TRIANGULATE_H
 
-#include "core/math/vector2.h"
-#include "core/templates/vector.h"
+#include "core/math/math_defs.h"
+struct Vector2;
```
Comment (AThousandShips) [inline core/math/triangulate.h:35]:
```suggestion

struct Vector2;

```

File: core/math/triangulate.cpp [hunk 8]
```diff
@@ -29,6 +29,8 @@
 /**************************************************************************/
 
 #include "triangulate.h"
+#include "core/math/vector2.h"
```
Comment (AThousandShips) [inline core/math/triangulate.cpp:32]:
```suggestion

#include "core/math/vector2.h"
```

**Comment by Nazarwadim** [comment]
I removed unnecessary trivial inclusions like 
* core/typedefs.h
* core/error/error_macros.h
* core/os/memory.h
* core/math/math_defs.h

Also I removed a `variant`, `object`, when `ref_counted` is present. And also something like `vector_n.h` when there is a `variant`.

**Comment by AThousandShips** [comment]
Will take a new look but I'm not seeing the need or reason for the vast majority of this, so I'm not sure this should be made at all

**Pull request closed without merging**